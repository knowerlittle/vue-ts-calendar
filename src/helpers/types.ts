export type SelectedDate = {
  fromDate: string
  toDate: string
  label: string
}
