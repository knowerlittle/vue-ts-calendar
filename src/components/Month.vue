<template>
  <div :class="$style.month">

    <DayOfWeekNames
      v-if="showDayNames"
      :dayNames="dayNames"
    />

    <div :class="$style['month__navigation']">
      <button
        v-if="showPreviousMonth"
        :class="$style['month__navigation-left']"
        :disabled="disablePrevious"
        @click="changeMonth(offsetAmount * -1)"
      >
        Prev
      </button>

      <p
        :class="[
          $style['month__navigation-name'], {[
          $style['align-right']]: monthAlignRight
        }]"
      >
        {{ selectedMonth }} {{ selectedYear }}
      </p>

      <button
        v-if="showNextMonth"
        :class="$style['month__navigation-right']"
        :disabled="disableNext"
        @click="changeMonth(offsetAmount)"
      >
        Next
      </button>
    </div>

    <div :class="[$style['month__days'], {[$style['month__days-wrapper']]: !isInfinite}]">
      <Day
        v-for="date in dates"
        :key="`${selectedYear}-${selectingMonth}-${date.date}`"
        :firstSaturdayDate="firstSaturdayDate"
        :date="date"
        @click="selectDate"
      />
    </div>

  </div>
</template>

<script lang="ts">
import { DateTime, Interval } from 'luxon'
import { Component, Vue, Prop, Inject } from 'vue-property-decorator'

import DayOfWeekNames from '@/components/DayOfWeekNames.vue'
import Day from '@/components/Day.vue'
import CalendarDate from '@/services/CalendarDate'

@Component({
  name: 'Month',
  components: {
    Day,
    DayOfWeekNames
  }
})
export default class Month extends Vue {
  @Inject() readonly currentDate!: DateTime

  /**
   * Show/Hide day names; Used when day names are external and sticky
   */
  @Prop({ default: true })
  readonly showDayNames!: boolean

  /**
   * Current range selection
   */
  @Prop()
  readonly currentSelectionRange: Interval | undefined

  /**
   * Maximum range selection
   */
  @Prop({ required: true })
  readonly maximumSelectionRange!: Interval

  /**
   * Show previous month button
   */
  @Prop({ required: true })
  readonly showPreviousMonth!: boolean

  /**
   * Show next month button
   */
  @Prop({ required: true })
  readonly showNextMonth!: boolean

  /**
   * Amount of offset emitted when a month changes
   */
  @Prop({ default: 1 })
  readonly offsetAmount!: number

  /**
   * The offset base on the month from the current date that the calendar will display
   */
  @Prop({ required: true })
  readonly monthOffset!: number

  /**
   * Sets the month name alignment to the right
   */
  @Prop({ default: false })
  readonly monthAlignRight!: boolean

  /**
   * Name of months
   */
  @Prop({ required: true })
  readonly monthNames!: string[]

  /**
   * Name of days
   */
  @Prop({ required: true })
  readonly dayNames!: string[]

  /**
   * If it is infinite scroll
   */
  @Prop({ default: false })
  readonly isInfinite!: boolean

  $style: any

  monthChange = 0

  get offSetMonth () {
    return this.monthChange + this.monthOffset
  }

  get viewingDate () {
    return this.currentDate.plus({ months: this.offSetMonth })
  }

  get selectingMonth () {
    return this.viewingDate.month - 1
  }

  get selectedYear () {
    return this.viewingDate.year
  }

  get selectedMonth () {
    return this.monthNames[this.selectingMonth]
  }

  get selectingMonthOneOffset () {
    return this.selectingMonth + 1
  }

  // date + 1 is required as date starts from 0 when we want a calendar date
  get dates () {
    return new Array(this.viewingDate.daysInMonth).fill({}).map((key, date) => {
      return new CalendarDate(
        date + 1,
        this.selectingMonthOneOffset,
        this.selectedYear,
        this.currentSelectionRange,
        this.maximumSelectionRange
      )
    })
  }

  get firstSaturdayDate () {
    return 8 - this.dates[0].weekDay()
  }

  get disablePrevious () {
    return this.maximumSelectionRange.isAfter(this.viewingDate.plus({ months: -1 }))
  }

  get disableNext () {
    const monthDifference = this.maximumSelectionRange.end.day > this.viewingDate.day
      ? 1
      : 0
    return this.maximumSelectionRange.isBefore(this.viewingDate.plus({ months: monthDifference }))
  }

  /**
   * Emits the amount of month to change by
   * @type {number}
   */
  private changeMonth (value: number) {
    this.$emit('changeMonth', value)
  }

  /**
   * Emits selected date
   * @type {CalendarDate}
   */
  private selectDate (date: CalendarDate) {
    this.$emit('selectDate', date)
  }
}
</script>

<style lang="scss" module>
  $month-font-size: 20px;

  .month {
    box-sizing: border-box;
    padding: 1px;
    width: 100%;

    p {
      justify-self: center;
    }

    &__navigation {
      display: grid;
      grid-template-columns: repeat(7, 1fr);
      padding: 8px 0 4px 0;
    }

    &__navigation-left {
      grid-column: 1 / 2;
      align-self: center;
      justify-self: center;
      cursor: pointer;
    }

    &__navigation-right {
      grid-column: 7 / 8;
      align-self: center;
      justify-self: center;
      cursor: pointer;
    }

    &__navigation-name {
      font-size: $month-font-size;
      color: black;
      grid-column: 2 / 7;

      &.align-right {
        grid-column: 1 / 8;
        justify-self: end;
        margin-right: 5%
      }
    }

    &__days {
      display: grid;
      grid-row-gap: 0.5rem;
    }

    &__days-wrapper {
      grid-template-rows: repeat(6, 1fr)
    }
  }

</style>
