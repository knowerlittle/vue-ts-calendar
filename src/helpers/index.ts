export const monthNamesFallback = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December'
]

export const dayNamesFallback = [
  'Mon',
  'Tues',
  'Wed',
  'Thu',
  'Fri',
  'Sat',
  'Sun'
]

export enum ViewType {
  Single = 'single',
  Dual = 'dual',
  Infinite = 'infinite'
}
