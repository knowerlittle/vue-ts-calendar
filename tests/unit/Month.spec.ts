import { shallowMount } from '@vue/test-utils'
import { DateTime, Interval } from 'luxon'

import Month from '@/components/Month.vue'
import { dayNamesFallback, monthNamesFallback } from '@/helpers'

const mockIsAfter = jest.fn()
const mockIsBefore = jest.fn()

// TODO: extract out to a seperate __mocks__ folders; couldn't figure out how
jest.mock('luxon', () => {
  const mockLuxon = {
    plus: jest.fn().mockReturnValue({
      local: jest.fn(),
      fromISO: jest.fn(),
      month: 2,
      year: 2020,
      plus: jest.fn()
    })
  }

  return {
    DateTime: {
      plus: jest.fn(),
      local: jest.fn().mockReturnValue(mockLuxon),
      fromISO: jest.fn().mockReturnValue(mockLuxon)
    },
    Interval: {
      hasSame: jest.fn(),
      contains: jest.fn(),
      plus: jest.fn(),
      fromDateTimes: () => ({
        fromDateTimes: jest.fn(),
        hasSame: jest.fn(),
        contains: jest.fn(),
        isAfter: mockIsAfter,
        isBefore: mockIsBefore,
        plus: jest.fn(),
        end: {
          day: jest.fn()
        }
      })
    }
  }
})

const mockDateTime = DateTime.local()

describe('Month', () => {
  const defaultOpts = {
    propsData: {
      maximumSelectionRange: Interval.fromDateTimes(mockDateTime, mockDateTime),
      showDayNames: true,
      showPreviousMonth: true,
      showNextMonth: true,
      offsetAmount: 1,
      monthOffset: 2,
      monthAlignRight: false,
      dayNames: dayNamesFallback,
      monthNames: monthNamesFallback
    },
    provide: {
      currentDate: DateTime.local(),
      isInfinite: false
    }
  }

  const factory = (opts = {}) => shallowMount(Month, {
    ...defaultOpts,
    ...opts
  })

  describe('matches snapshot', () => {
    it('with default cases', () => {
      const wrapper = factory()

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('when there are no names of weeks', () => {
      const options = {
        propsData: {
          maximumSelectionRange: Interval.fromDateTimes(mockDateTime, mockDateTime),
          showDayNames: false,
          showPreviousMonth: true,
          showNextMonth: true,
          monthOffset: 2,
          dayNames: dayNamesFallback,
          monthNames: monthNamesFallback
        }
      }

      const wrapper = factory(options)

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('when month align rights and there are no month navigations', () => {
      const options = {
        propsData: {
          maximumSelectionRange: Interval.fromDateTimes(mockDateTime, mockDateTime),
          monthAlignRight: true,
          showDayNames: true,
          showPreviousMonth: false,
          showNextMonth: false,
          monthOffset: 2,
          dayNames: dayNamesFallback,
          monthNames: monthNamesFallback
        }
      }

      const wrapper = factory(options)

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('when both month navigations are disabled', () => {
      mockIsBefore.mockReturnValueOnce(true)
      mockIsAfter.mockReturnValueOnce(true)

      const wrapper = factory()

      expect(wrapper.html()).toMatchSnapshot()
    })

    it('when isInfinite is true', () => {
      const wrapper = factory({
        provide: {
          currentDate: DateTime.local(),
          isInfinite: true
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('events', () => {
    it('should emit an event to change the month to previous month when clicked', async () => {
      mockIsAfter.mockReturnValueOnce(false)
      const wrapper = factory()

      wrapper.find('.month__navigation-left').trigger('click')
      await wrapper.vm.$nextTick()

      // @ts-ignore
      expect(wrapper.emitted('changeMonth')[0][0]).toEqual(-1)
    })

    it('should emit an event to change the month to next month when clicked', async () => {
      mockIsBefore.mockReturnValueOnce(false)
      const wrapper = factory()

      wrapper.find('.month__navigation-right').trigger('click')
      await wrapper.vm.$nextTick()

      // @ts-ignore
      expect(wrapper.emitted('changeMonth')[0][0]).toEqual(1)
    })

    it('should emit a date object when clicked', () => {
      const wrapper = factory()

      const dateObject = {
        date: '2020/12/12'
      }
      wrapper.find('day-stub').vm.$emit('click', dateObject)

      // @ts-ignore
      expect(wrapper.emitted('selectDate')[0][0]).toEqual(dateObject)
    })
  })
})
