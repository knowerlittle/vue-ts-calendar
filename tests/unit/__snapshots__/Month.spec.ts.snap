// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Month matches snapshot when both month navigations are disabled 1`] = `
<div class="month">
  <dayofweeknames-stub daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun"></dayofweeknames-stub>
  <div class="month__navigation"><button disabled="disabled" class="month__navigation-left">
      Prev
    </button>
    <p class="month__navigation-name">
      February 2020
    </p> <button disabled="disabled" class="month__navigation-right">
      Next
    </button>
  </div>
  <div class="month__days month__days-wrapper">
    <day-stub date="[object Object]" firstsaturdaydate="NaN"></day-stub>
  </div>
</div>
`;

exports[`Month matches snapshot when isInfinite is true 1`] = `
<div class="month">
  <dayofweeknames-stub daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun"></dayofweeknames-stub>
  <div class="month__navigation"><button class="month__navigation-left">
      Prev
    </button>
    <p class="month__navigation-name">
      February 2020
    </p> <button class="month__navigation-right">
      Next
    </button>
  </div>
  <div class="month__days month__days-wrapper">
    <day-stub date="[object Object]" firstsaturdaydate="NaN"></day-stub>
  </div>
</div>
`;

exports[`Month matches snapshot when month align rights and there are no month navigations 1`] = `
<div class="month">
  <dayofweeknames-stub daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun"></dayofweeknames-stub>
  <div class="month__navigation">
    <!---->
    <p class="month__navigation-name align-right">
      February 2020
    </p>
    <!---->
  </div>
  <div class="month__days month__days-wrapper">
    <day-stub date="[object Object]" firstsaturdaydate="NaN"></day-stub>
  </div>
</div>
`;

exports[`Month matches snapshot when there are no names of weeks 1`] = `
<div class="month">
  <!---->
  <div class="month__navigation"><button class="month__navigation-left">
      Prev
    </button>
    <p class="month__navigation-name">
      February 2020
    </p> <button class="month__navigation-right">
      Next
    </button>
  </div>
  <div class="month__days month__days-wrapper">
    <day-stub date="[object Object]" firstsaturdaydate="NaN"></day-stub>
  </div>
</div>
`;

exports[`Month matches snapshot with default cases 1`] = `
<div class="month">
  <dayofweeknames-stub daynames="Mon,Tues,Wed,Thu,Fri,Sat,Sun"></dayofweeknames-stub>
  <div class="month__navigation"><button class="month__navigation-left">
      Prev
    </button>
    <p class="month__navigation-name">
      February 2020
    </p> <button class="month__navigation-right">
      Next
    </button>
  </div>
  <div class="month__days month__days-wrapper">
    <day-stub date="[object Object]" firstsaturdaydate="NaN"></day-stub>
  </div>
</div>
`;
